From 2536b0f9285103ca32ed3d360e7026fc88d3547c Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 18 Oct 2023 15:20:21 +0530
Subject: [PATCH 4/4] hostapd: refresh other iface beacons when one iface gets
 disabled/enabled

If one or more bss from the iface is partnering with bsses from other
iface and if this iface get disabled, then the beacons needs to be
refreshed for other ifaces. Similar thing should happen when it gets
enabled.

Add logic to refresh other iface beacons when one of the iface is disabled
or enabled.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 src/ap/hostapd.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -57,6 +57,7 @@
 #include "wpa_auth_kay.h"
 #include "hw_features.h"
 
+static void hostapd_refresh_all_iface_beacons(struct hostapd_iface *hapd_iface);
 
 static int hostapd_flush_old_stations(struct hostapd_data *hapd, u16 reason);
 #ifdef CONFIG_WEP
@@ -3279,6 +3280,8 @@ int hostapd_enable_iface(struct hostapd_
 		return -1;
 	}
 
+	hostapd_refresh_all_iface_beacons(hapd_iface);
+
 	return 0;
 }
 
@@ -3318,6 +3321,22 @@ int hostapd_reload_bss_only(struct hosta
 }
 
 
+static void hostapd_refresh_all_iface_beacons(struct hostapd_iface *hapd_iface)
+{
+	size_t j;
+
+	if (!hapd_iface || !hapd_iface->interfaces ||
+	    hapd_iface->interfaces->count <= 1)
+		return;
+
+	for (j = 0; j < hapd_iface->interfaces->count; j++) {
+		if (hapd_iface->interfaces->iface[j] == hapd_iface)
+			continue;
+
+		ieee802_11_update_beacons(hapd_iface->interfaces->iface[j]);
+	}
+}
+
 int hostapd_disable_iface(struct hostapd_iface *hapd_iface)
 {
 	size_t j;
@@ -3369,6 +3388,9 @@ int hostapd_disable_iface(struct hostapd
 		   hapd_iface->bss[0]->conf->iface);
 	hostapd_set_state(hapd_iface, HAPD_IFACE_DISABLED);
 	hostapd_interface_update_fils_ubpr(hapd_iface, false);
+
+	hostapd_refresh_all_iface_beacons(hapd_iface);
+
 	return 0;
 }
 
